[SPARK-18274][ML][PYSPARK] Memory leak in PySpark JavaWrapper

techaddict opened the pull request:
## What changes were proposed in this pull request?
In`JavaWrapper `'s destructor make Java Gateway dereference object in destructor, using `SparkContext._active_spark_context._gateway.detach` 
Fixing the copying parameter bug, by moving the `copy` method from `JavaModel` to `JavaParams`

## How was this patch tested?
```scala
import random, string
from pyspark.ml.feature import StringIndexer

l = [(''.join(random.choice(string.ascii_uppercase) for _ in range(10)), ) for _ in range(int(7e5))]  # 700000 random strings of 10 characters
df = spark.createDataFrame(l, ['string'])

for i in range(50):
    indexer = StringIndexer(inputCol='string', outputCol='index')
    indexer.fit(df)
```
* Before: would keep StringIndexer strong reference, causing GC issues and is halted midway
After: garbage collection works as the object is dereferenced, and computation completes
* Mem footprint tested using profiler
* Added a parameter copy related test which was failing before.

techaddict committed:
[SPARK-18274] Memory leak in PySpark StringIndexer

SparkQA commented:
**[Test build #68478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68478/consoleFull)** for PR 15843 at commit [`a493c19`](https://github.com/apache/spark/commit/a493c1961829000986446db11ce67f3103a79bea).

techaddict committed:
change

SparkQA commented:
**[Test build #68482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68482/consoleFull)** for PR 15843 at commit [`f25b099`](https://github.com/apache/spark/commit/f25b0992f0a18177c84203419aca8f52eaf1c6df).

SparkQA commented:
**[Test build #68478 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68478/consoleFull)** for PR 15843 at commit [`a493c19`](https://github.com/apache/spark/commit/a493c1961829000986446db11ce67f3103a79bea).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68478/
Test FAILed.

SparkQA commented:
**[Test build #68482 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68482/consoleFull)** for PR 15843 at commit [`f25b099`](https://github.com/apache/spark/commit/f25b0992f0a18177c84203419aca8f52eaf1c6df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68482/
Test PASSed.

techaddict commented:
cc: @jkbradley @davies @holdenk

jkbradley commented:
Thanks a lot for finding & reporting this!  The fix should probably go in JavaWrapper, not JavaModel, right?

I tested this manually (in JavaWrapper), and it seems to fix the problematic case with StringIndexer.

techaddict commented:
@jkbradley yes I did it for `JavaWrapper` first, but try running tests with it gives https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68478/consoleFull

jkbradley commented:
You're right!  It's another bug: copy should be implemented in JavaParams, not JavaModel.  I'm sending this PR to fix that: https://github.com/techaddict/spark/pull/1

Can you please check it out and merge it into your PR if it looks OK to you?  All pyspark.ml tests ran successfully with it.

jkbradley committed:
Fixing copy bug (#1)

* moved copy from JavaModel to JavaParams. mv del from JavaModel to JavaWrapper

* added test which fails before this fix

techaddict commented:
@jkbradley looks good, merged 👍

SparkQA commented:
**[Test build #68513 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68513/consoleFull)** for PR 15843 at commit [`3d858a2`](https://github.com/apache/spark/commit/3d858a2326809b7e1c679b712d84a8a21767d13c).

techaddict committed:
check if there is active_spark_context

causes this error while quitting pyspark:

Exception ignored in: <bound method JavaWrapper.__del__ of
StringIndexer_4a75b9e8c92f56703aff>
Traceback (most recent call last):
  File "/Users/pichu/Project/Spark/python/pyspark/ml/wrapper.py", line
37, in __del__
    SparkContext._active_spark_context._gateway.detach(self._java_obj)
AttributeError: 'NoneType' object has no attribute '_gateway'
Exception ignored in: <bound method JavaWrapper.__del__ of
StringIndexer_4a75b9e8c92f56703aff>
Traceback (most recent call last):
  File "/Users/pichu/Project/Spark/python/pyspark/ml/wrapper.py", line
37, in __del__
AttributeError: 'NoneType' object has no attribute '_gateway'

techaddict committed:
Merge branch 'master' into SPARK-18274

techaddict commented on python/pyspark/ml/wrapper.py:
```diff
@@ -33,6 +33,10 @@ def __init__(self, java_obj=None):
         self._java_obj = java_obj
 
+    def __del__(self):
+        if SparkContext._active_spark_context:
```
checking if there is active spark context, got this error after `quit()` in `pyspark`
```
Exception ignored in: <bound method JavaWrapper.__del__ of
StringIndexer_4a75b9e8c92f56703aff>
Traceback (most recent call last):
File "/Users/xx/Project/Spark/python/pyspark/ml/wrapper.py", line
37, in __del__
SparkContext._active_spark_context._gateway.detach(self._java_obj)
AttributeError: 'NoneType' object has no attribute '_gateway'
Exception ignored in: <bound method JavaWrapper.__del__ of
StringIndexer_4a75b9e8c92f56703aff>
Traceback (most recent call last):
File "/Users/xx/Project/Spark/python/pyspark/ml/wrapper.py", line
37, in __del__
AttributeError: 'NoneType' object has no attribute '_gateway'
```

SparkQA commented:
**[Test build #68514 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68514/consoleFull)** for PR 15843 at commit [`dc5aee3`](https://github.com/apache/spark/commit/dc5aee399707f712bd4627dc41719760a0db997d).

SparkQA commented:
**[Test build #68513 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68513/consoleFull)** for PR 15843 at commit [`3d858a2`](https://github.com/apache/spark/commit/3d858a2326809b7e1c679b712d84a8a21767d13c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68513/
Test PASSed.

SparkQA commented:
**[Test build #68514 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68514/consoleFull)** for PR 15843 at commit [`dc5aee3`](https://github.com/apache/spark/commit/dc5aee399707f712bd4627dc41719760a0db997d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `final class ParquetLogRedirector implements Serializable `
  * `  case class OutputSpec(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68514/
Test PASSed.

viirya commented on python/pyspark/ml/wrapper.py:
```diff
@@ -180,6 +184,24 @@ def __get_class(clazz):
+        """
+        Creates a copy of this instance with the same uid and some
+        extra params. This implementation first calls Params.copy and
+        then make a copy of the companion Java model with extra params.
```
nit: Java model -> Java pipeline component.

techaddict committed:
nit: doc fix

viirya commented on python/pyspark/ml/wrapper.py:
```diff
@@ -180,6 +184,25 @@ def __get_class(clazz):
+        Creates a copy of this instance with the same uid and some
+        extra params. This implementation first calls Params.copy and
+        then make a copy of the companion Java pipeline component with
+        extra params. So both the Python wrapper and the Java model get
```
nit: There is another "Java model" here too.

viirya commented:
LGTM with minor doc comment.

SparkQA commented:
**[Test build #68515 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68515/consoleFull)** for PR 15843 at commit [`01a80b9`](https://github.com/apache/spark/commit/01a80b9d783dca7e74b717b0374305cb4376208a).

techaddict committed:
doc

SparkQA commented:
**[Test build #68516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68516/consoleFull)** for PR 15843 at commit [`a76a1fb`](https://github.com/apache/spark/commit/a76a1fb1f10532e0e99592c53fbaa548279f69a8).

SparkQA commented:
**[Test build #68515 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68515/consoleFull)** for PR 15843 at commit [`01a80b9`](https://github.com/apache/spark/commit/01a80b9d783dca7e74b717b0374305cb4376208a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68515/
Test PASSed.

SparkQA commented:
**[Test build #68516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68516/consoleFull)** for PR 15843 at commit [`a76a1fb`](https://github.com/apache/spark/commit/a76a1fb1f10532e0e99592c53fbaa548279f69a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68516/
Test PASSed.

holdenk commented:
So this change looks good to me, but it seems like it fixes more than just the bug described in the JIRA & PR description with @jkbradley's change integrated (namely the issue with param copy which we have). For people who are looking for what's changed between versions it might make sense to explain the copy related fix the PR description as well since that is what is used in the commit log.

techaddict committed:
Merge branch 'master' into SPARK-18274

techaddict commented:
@holdenk updated the description.

holdenk commented:
LGTM thanks for fixing this @techaddict :D :)

holdenk commented:
ping @davies if you have time for final review/merge?

jkbradley commented:
Good point @holdenk --- @techaddict could you also please update the PR title to say "JavaWrapper" instead of "StringIndexer"?

jkbradley commented:
I'm now wondering if ```__del___``` should be in JavaParams instead of JavaWrapper since JavaWrapper does not override ```copy```.  Do yall agree it will be safer if it's moved there?

viirya commented:
@jkbradley Sounds making sense more.

holdenk commented:
I'm not so sure about that, we still would want to cleanup the underlying Java reference object on delete if it isn't needed anymore. I think the question is do we want to support shallow copy of javawrapper objects?

viirya commented:
Oh. I thought JavaWrapper is only used on JavaParams. But there are also others like LogisticRegressionSummary which directly inherits JavaWrapper.

jkbradley commented:
I don't see a need to do deep copies of model summaries, but I agree I don't like how JavaWrapper is ambiguous about whether it does shallow or deep copies.

I'd say the confusion comes from us having a mix of immutable Java types (like model summaries) and mutable Java types (like Params subclasses).  What do you think of these 2 options?
* Distinguish mutability within Python wrappers: JavaWrapper is usable for immutable types.  JavaParams (or other subtypes, if needed) is usable for mutable types.  I.e., ```__del__``` and ```copy``` go in JavaParams.
* Distinguish mutability within Java only: Use the same wrapper types for both in Python, and Java copy methods can do deep or shallow copies.  I.e., in JavaWrapper, implement copy() which copies the Java instance, and implement ```__del__``` to release that instance's handle.

I don't think either option does much for enforcing these semantics.  Barring GC issues, I'd pick option 1 since it's simpler.  But if option 2 is better for GC issues, then I'd vote for it.

Thoughts?

viirya commented:
I'd prefer option2 for safety since the model summaries should be an issue for GC. And looks like Java model summaries don't have copy method.

holdenk commented:
From the Py4J documentation it seems like we could be leaking memory with the first option, although perhaps not a lot of memory, but if it was being used in an iterative Python algorithm for training many models it could start to have some impact. I'd be in favor of option 2, but that could be done as a follow up issue if the required copy methods aren't generally available.

viirya commented:
@jkbradley @holdenk @techaddict Do we want to implement copy() in JavaWrapper in this PR too? Or separate it to follow ones with the required copy methods on JVM side?

jkbradley commented:
Sorry for the slow response on this.  Given the time pressure for 2.1, let's go with option 1 for now with a follow-up task to implement option 2.  It would be great to include this fix in 2.1.

@techaddict will you have time to update your PR quickly?  Thank you!

holdenk commented:
I agree, for a follow up (so we don't lose track of it) - I've created SPARK-18630 but option 1 for now is a strict improvement over the current situation. Thanks for all of your work on this @techaddict

techaddict commented:
@jkbradley @holdenk will update the PR with changes today.

techaddict committed:
Merge branch 'master' into SPARK-18274

techaddict committed:
move del to javaparams

SparkQA commented:
**[Test build #69476 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69476/consoleFull)** for PR 15843 at commit [`37e83e8`](https://github.com/apache/spark/commit/37e83e88fe84dcfdcd04426c85499f7494cd688b).

SparkQA commented:
**[Test build #69476 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69476/consoleFull)** for PR 15843 at commit [`37e83e8`](https://github.com/apache/spark/commit/37e83e88fe84dcfdcd04426c85499f7494cd688b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69476/
Test PASSed.

techaddict commented:
@jkbradley @holdenk @viirya PR updated

viirya commented:
LGTM

jkbradley commented:
LGTM too
Thanks a lot!
Merging with master, branch-2.1, branch-2.0

Has anyone heard of complaints of this in current use cases of earlier branches?  If not, I won't backport it further than 2.0.
